fix(forms): handle form groups/arrays own pending async validation

benbraou opened the pull request:
introduce a boolean to track form groups/arrays own pending async validation to distinguish between pending state due to children and pending state due to own validation

Fixes #10064

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
**Scenario**:  

1.  An async validator is set up on a `form control/ array`
1. Async validators are set up on child form controls
1. All of the child controls validation is successful but the `form control/ array` validator has not resolved yet

Currently, the `form control/array` is considered valid even if its own validator has not resolved yet

Issue Number: N/A
#10064

## What is the new behavior?
The `form control/array` is in pending state if its own validator has not resolved yet and none of its children is invalid

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -590,12 +617,31 @@ export abstract class AbstractControl {
+   * @param hasSomeInvalidControls Whether some of the control children have INVALID status
+   * @returns Whether a group of controls is in PENDING state due to its own validation
+   */
+  private _isPendingDueToOwnValidation(hasSomeInvalidControls: boolean): boolean {
```
It looks like passing the `hasSomeInvalidControls` argument can be avoided (it can be included into conditions where the `_isPendingDueToOwnValidation` function is being called).

benbraou replied:
Indeed, this is unnecessarily complicated :smiley:
I am removing it

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -590,12 +617,31 @@ export abstract class AbstractControl {
     if (this.errors) return INVALID;
-    if (this._anyControlsHaveStatus(PENDING)) return PENDING;
-    if (this._anyControlsHaveStatus(INVALID)) return INVALID;
+
```
The code below changes the logic in some backwards-incompatible way: prior to this change 1 pending control + 1 invalid control resulted in the "pending" state. After the change in such scenario the state would be calculated as "invalid". Would that be enough to update the "pending" state calculation to include the check for own async validator state as shown below?

```
    if (this._isPendingDueToOwnValidation() || this._anyControlsHaveStatus(PENDING)) return PENDING;
```

benbraou replied:
Yes, let's keep the previous behaviour.I will change this. Thank you.
I was thinking that since the form is guaranteed to be invalid after all pending validators resolve (since one of them failed), we could already set to invalid even if there are validators pending.

AndrewKushnir reviewed (commented):
Thanks for creating this PR @benbraou 👍 
It looks like the change is in the right direction and I've added a couple comments.

benbraou committed:
fix(forms): handle form groups/arrays own pending async validation

introduce a boolean to track form groups/arrays own pending async validation to distinguish between pending state due to children and pending state due to own validation

Fixes #10064

AndrewKushnir commented:
Hi @benbraou, thanks for the reply and applying additional changes 👍 

I had another look and I'm wondering if we can reuse existing `_asyncValidationSubscription` field as an indication that async validator is "in progress"? I.e. we can probably set it to `undefined` once async validator is completed (and may be have a function like `_hasOwnPendingAsyncValidator` that would return `boolean` based on the value of the `_asyncValidationSubscription` field). What do you think about this solution?

Also there are some failing tests on CI, could you please look into them? It looks like most of them are added as a part of this PR.

FYI, I also started tests in Google's codebase ([blueprint](http://test/OCL:316303860:BASE:316303868:1592104886793:d7f386ad) and [global](https://test.corp.google.com/OCL:316303860:BASE:316303957:1592105009293:d3ba15e1), links are internal-only) to see how this change might affect existing apps.

Thank you.

benbraou commented:
Hi @AndrewKushnir. Thank you for the review. Sorry about the failing tests, I should've let you know that I haven't finished updating them yet. They should be fine now.
I am now using only `_asyncValidationSubscription ` as you suggested.
```typescript
      this._asyncValidationSubscription = obs.subscribe((errors: ValidationErrors|null) => {
        this._asyncValidationSubscription = undefined;
        this.setErrors(errors, {emitEvent});
      });
```
I just have one comment, the order is important. The subscription needs to be set to undefined before setting the errors. This is because `this.setErrors ` will trigger a recalculation of the control status
```typescript
  setErrors(errors: ValidationErrors|null, opts: {emitEvent?: boolean} = {}): void {
    (this as {errors: ValidationErrors | null}).errors = errors;
    this._updateControlsErrors(opts.emitEvent !== false);
  }

  _updateControlsErrors(emitEvent: boolean): void {
    (this as {status: string}).status = this._calculateStatus();
  ...
```
And `_calculateStatus` is now also based on whether `this._asyncValidationSubscription` is undefined.
Not sure, if this is error prone in the future and whether a comment is needed?
Thank you

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1008,313 @@ describe('FormGroup', () => {
+       fakeAsync(() => {
+         const c = new FormControl('fcValue', null!, asyncValidator('fcValue', {fcValue: 1}));
+         const g = new FormGroup(
+             {'one': c}, null!, asyncGroupValidator({timeout: 2, shouldThrowError: true}));
```
I really like the new validator, which has just the necessary logic, i.e. timeout and whether validation should be successful or not 👍 It'd be great to use the same validator for the `FormControl` as well, so it's consistent:

```
const c = new FormControl('fcValue', null!, simpleAsyncValidator({timeout: 10, shouldFail: false}));
const g = new FormGroup({'one': c}, null!, simpleAsyncValidator({timeout: 20, shouldFail: true}));
```

benbraou replied:
Sounds good. I am now using `asyncValidator ` for all the tests that I added.
Please let me know if I also should refactor the existing tests to use the new validator as well. Maybe this can be done in a separate commit or separate PR, or just wait :)

AndrewKushnir replied:
Let's make a refactor in a separate PR. I believe we may also need to look into combining `asyncValidator` and `simpleAsyncValidator` in tests.

benbraou replied:
That makes sense. Thank you!

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1008,313 @@ describe('FormGroup', () => {
+
+         tick(1);
+
+         // After 1s, only form control validation has resolved
```
Minor comment: since you use the timeout value in `setTimeout` as is, the timeout is in `ms`.

benbraou replied:
Indeed. Thank you

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -676,8 +687,10 @@ export abstract class AbstractControl {
-      this._asyncValidationSubscription =
-          obs.subscribe((errors: ValidationErrors|null) => this.setErrors(errors, {emitEvent}));
+      this._asyncValidationSubscription = obs.subscribe((errors: ValidationErrors|null) => {
+        this._asyncValidationSubscription = undefined;
```
> Not sure, if this is error prone in the future and whether a comment is needed?

Yeah, I think adding a comment here would be great, thanks.

AndrewKushnir replied:
After looking at this a bit more, I think we should probably:
- call `this._cancelExistingSubscription()` here
- update the `_cancelExistingSubscription` method to set `this._asyncValidationSubscription` to `undefined` after unsubscribe operation.

What do you think?

benbraou replied:
That  makes sense. Thank you

AndrewKushnir reviewed (changes requested):
Thanks for the updates @benbraou 👍

I've left a few comments, please have a look.

Also I wanted to mention that I ran tests in Google's codebase and there are few cases where additional investigation is needed. I will keep you updated.

Thank you.

benbraou commented on packages/forms/src/model.ts:
```diff
@@ -137,6 +137,17 @@ export abstract class AbstractControl {
+   * Indicates that a control has its own pending asynchronous validation that runs in addition to
+   * the validations occurring in its child controls.
+   *
+   * @since #10064
```
I ll remove it. Thank you

benbraou commented on packages/forms/src/model.ts:
```diff
@@ -137,6 +137,17 @@ export abstract class AbstractControl {
   _pendingDirty!: boolean;
 
+  /**
+   * Indicates that a control has its own pending asynchronous validation that runs in addition to
```
I ll update it. Thank you

benbraou commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1008,313 @@ describe('FormGroup', () => {
+
+         expect(g.get('one')!.errors).toEqual(null);
+         expect(g.get('one')!.pending).toEqual(true);
+         expect(g.get('one')!.status).toEqual('PENDING');
```
Sounds great! Initially, I was creating of creating a custom matcher, but I did not see this being used in the project.
Yes, `currentStateOf ` makes the test much more readable.
I think handling both FormControl or list of FormControl as parameters is useful
```typescript
interface ControlValidationStatus {
  errors: any;
  pending: boolean;
  status: string;
}

function currentStateOf<T extends AbstractControl|AbstractControl[]>(c: T):
    T extends AbstractControl ? ControlValidationStatus : ControlValidationStatus[];

function currentStateOf<T extends AbstractControl|AbstractControl[]>(c: T): ControlValidationStatus|
    ControlValidationStatus[] {
  const singleFormControlStatus = (fc: AbstractControl) => {
    const {errors, pending, status} = fc;
    return ({errors, pending, status});
  };

  if (Array.isArray(c)) {
    return c.map(singleFormControlStatus);
  }

  return singleFormControlStatus(c as AbstractControl);
}
```

I am pushing this change in minutes, but please let me know if I should only support list of FormControl as parameters.
Thank you for the review and your time 🙏

AndrewKushnir replied:
Thanks for making this change 👍 Test results are very descriptive and intuitive now!

> please let me know if I should only support list of FormControl as parameters.

Yeah, I think we should simplify that logic and typings and just take care of list of Controls (remove the support for getting the status of a single control).

benbraou replied:
Great. I am pushing the change in a couple of minutes. Thank you

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -2074,6 +2074,60 @@ import {MyInput, MyInputForm} from './value_accessor_integration_spec';
+
+           tick(100);
+
+           // login form control is valid. However, the form control is invalid because `angul` does
```
```suggestion
           // Login form control is valid. However, the form control is invalid because `angul` does
```

benbraou replied:
Great. I am pushing the change in a couple of minutes. Thank you

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -2074,6 +2074,60 @@ import {MyInput, MyInputForm} from './value_accessor_integration_spec';
+           expect(form.pending).toEqual(false);
+           expect(form.invalid).toEqual(true);
+
+           input.nativeElement.value = 'angular!';
```
I really like how you added comments above different steps in the `form_group_spec.ts` file. I'd propose having similar comments in this test as well, for example:

```
  // Setting a value that would be trigger "VALID" form state
  input.nativeElement.value = 'angular!';
  ...
```

and in a couple places below.

benbraou replied:
I have added the necessary comments and will be pushing in minutes. Thank you!

AndrewKushnir reviewed (commented):
The change looks good, thanks @benbraou 👍 

I left a couple comments, please have a look when you get a chance. I'd also propose adding some tests to cover basic cases of the `FormArray` class (with `FormControl`s and `FormGroup`s), what do you think?

Also a quick update: tests in Google's codebase went well. I'll re-run them again once all the changes are finalized.

benbraou commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1031,271 @@ describe('FormGroup', () => {
+
+         tick(1);
+
+         // After 1ms, form group validation is still pending, because child validation is pending
```
@AndrewKushnir fyi, here the assertion are correct, but the comment (`// After 1ms, form group validation is still pending, because child validation is pending`) is not.
Actually, `After 1ms, form group validation fails`, because the logic for calculating the status in `model.ts`, returns `INVALID` as soon as there are errors. Even if the form group has children with validation still pending
```typescript
  private _calculateStatus(): string {
    if (this._allControlsDisabled()) return DISABLED;
    if (this.errors) return INVALID;
    if (this._hasOwnPendingAsyncValidator() || this._anyControlsHaveStatus(PENDING)) return PENDING;
    if (this._anyControlsHaveStatus(INVALID)) return INVALID;
    return VALID;
  }
```
I am updating this comment to `// After 1ms, form group validation fails`

benbraou reviewed (commented):
> The change looks good, thanks @benbraou 👍
> 
> I left a couple comments, please have a look when you get a chance. I'd also propose adding some tests to cover basic cases of the `FormArray` class (with `FormControl`s and `FormGroup`s), what do you think?
> 
> Also a quick update: tests in Google's codebase went well. I'll re-run them again once all the changes are finalized.

I have added new tests for the FormArray. Thank you for the review 🙏 !

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1013,419 @@ describe('FormGroup', () => {
+         const g =
+             new FormGroup({'one': c}, null!, simpleAsyncValidator({timeout: 2, shouldFail: true}));
+
+         // Initially, the from group and nested control are in pending state
```
There are few places with the same typo, could you please have a look?
```suggestion
         // Initially, the form group and nested control are in pending state
```

benbraou replied:
I have updated all the occurrences of this typo. Thank you

AndrewKushnir reviewed (commented):
Thanks for additional tests @benbraou 👍 

I've just left a minor comment (there is a typo in test comments).

I will re-run tests in Google's codebase ([blueprint](http://test/OCL:317218248:BASE:317218268:1592527650135:94d8dc83) + [global](https://test.corp.google.com/ui#id=OCL:317218248:BASE:317228197:1592532090183:714a4fb8)) and let you know the result as soon as I have it.

Thank you.

benbraou commented:
@AndrewKushnir thank you for the continuous review and help. I have fixed the typos

AndrewKushnir commented:
Hi @benbraou, thanks for the fixes 👍 

Tests in Google's codebase resulted in a couple failures. After further investigation it turned out that my proposal to reuse the `_asyncValidationSubscription` field  was incorrect :( The thing is that some validators may return just `of(null)` (instead of more complicated setTimeout/Promise logic that we have in tests), which will synchronously emit values when we subscribe:
```
this._asyncValidationSubscription = obs.subscribe((errors: ValidationErrors|null) => {
  this._cancelExistingSubscription();
  this.setErrors(errors, {emitEvent});
});
```
So what'd happen in this case is that the callback function would be invoked synchronously and at that time the `this._asyncValidationSubscription` is not yet defined. As a part of the callback execution, we'll try to clear the subscription (which is not yet defined at this point) and apply the changes to the control and parent. Once callback function completes (and additional sync machinery in RxJS), the `this._asyncValidationSubscription` becomes defined. The problem is that the actual state is that validator already executed, but the `this._asyncValidationSubscription` field is still defined on a control, indicating that it's still pending.

So I think we should go back to your original approach of having a boolean flag (we can reuse the `_hasOwnPendingAsyncValidator` field and make it a boolean instead of the function) and revert other changes made to the `_cancelExistingSubscription` function. It'd be also great to add a few tests to illustrate the use case of a simple validator, i.e. when validator looks like this:
```
const validator = (c: AbstractControl) => of(null);
```

We can probably include this logic into the `simpleAsyncValidator` function that you developed and invoke that when `timeout === 0`. What do you think?

Sorry that my proposal turned out to be incorrect (thus we have to make this extra iteration) and thanks for all your help with this PR! Please let me know if any additional information is required.

Thank you.

benbraou commented:
@AndrewKushnir No problem, Thank you! I really appreciate the detailed explanations throughout the PR. That's very educational 👍 
I am now using the boolean flag to handle async validation, reverted the changes for canceling the async subscription, updated the `simpleAsyncValidator` to return `of(result)` if `timeout === 0`.
Also, I added a couple of tests.
I have just one remark regarding this new test that I added:
```typescript
    it('should handle failing async FormGroup and failing child control validators resolving synchronously',
       fakeAsync(() => {
         const c = new FormControl(
             'fcValue', null!, simpleAsyncValidator({timeout: 0, shouldFail: true}));
         const g =
             new FormGroup({'one': c}, null!, simpleAsyncValidator({timeout: 0, shouldFail: true}));

         // FormControl async validator has executed and failed synchronously with the deafult error
         // `{async: true}`. Next, the form group status is calculated. Since one of its children is
         // failing, the form group itself is marked `INVALID`. And its asynchronous validation is
         // not even triggered. Therefore, we end up with form group that is `INVALID` but whose
         // errors are null(child error do not propagate and own async validation not event
         // triggered)
         expect(currentStateOf([g, g.get('one')!])).toEqual([
           {errors: null, pending: false, status: 'INVALID'},           // Group
           {errors: {async: true}, pending: false, status: 'INVALID'},  // Control
         ]);
       }));
```
The comments should be clear, but basically, I was expecting the form group to have its own errors in its status.
However due to this logic, the async validator is not even run(the form group is `INVALID` because of its child)
```typescript
  updateValueAndValidity(opts: {onlySelf?: boolean, emitEvent?: boolean} = {}): void {
     ....
      if (this.status === VALID || this.status === PENDING) {
        this._runAsyncValidator(opts.emitEvent);
      }
     ...
    }
```
Anyway, if this behaviour should be changed, I think this is not linked to the original issue that this PR is solving
Thank you

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
          expect(g.get('one')!.errors).toEqual({'async': true});
        }));
+
+    it('should handle successful async FormGroup resolving synchrounously before a successful async child validator',
```
```suggestion
    it('should handle successful async FormGroup resolving synchronously before a successful async child validator',
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
+         ]);
+       }));
+
+    it('should handle successful async FormGroup resolving after a synchrounously and successfully resolving child validator',
```
```suggestion
    it('should handle successful async FormGroup resolving after a synchronously and successfully resolving child validator',
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
+         // `{async: true}`. Next, the form group status is calculated. Since one of its children is
+         // failing, the form group itself is marked `INVALID`. And its asynchronous validation is
+         // not even triggered. Therefore, we end up with form group that is `INVALID` but whose
+         // errors are null(child error do not propagate and own async validation not event
+         // triggered)
```
```suggestion
         // errors are null (child error do not propagate and own async validation not event
         // triggered).
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
+           {errors: null, pending: true, status: 'PENDING'},  // c2
+         ]);
+
+         tick(1);
+         // After 1ms, c2 validation has resolved
```
```suggestion
         tick(1);
         
         // After 1ms, c2 validation has resolved
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
+           {errors: null, pending: false, status: 'VALID'},   // c2
+         ]);
+
+         tick(1);
+         // After 1ms, FormArray own validation has failed
```
```suggestion
         tick(1);
         
         // After 1ms, FormArray own validation has failed
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
+           {errors: null, pending: false, status: 'VALID'},   // Control
+         ]);
+
+         tick(1);
+         // After 1ms, the form group validation resolves
```
```suggestion
         tick(1);
         
         // After 1ms, the form group validation resolves
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
+           {errors: null, pending: false, status: 'VALID'},   // Control
+         ]);
+
+         tick(1);
+         // After 1ms, one form async validator has resolved but not the second
```
```suggestion
         tick(1);
         
         // After 1ms, one form async validator has resolved but not the second
```

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
+           {errors: null, pending: true, status: 'PENDING'},  // Group
+           {errors: null, pending: false, status: 'VALID'},   // Control
+         ]);
+
+
+         tick(1);
```
```suggestion

         tick(1);
```
(removed extra line)

AndrewKushnir commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -981,6 +1014,531 @@ describe('FormGroup', () => {
+           {errors: null, pending: false, status: 'VALID'},             // Group g2
+         ]);
+
+         tick(1);
+         // After 1ms, g validation fails because g1 is invalid, but since errors do not cascade, so
```
```suggestion
         tick(1);
         
         // After 1ms, g validation fails because g1 is invalid, but since errors do not cascade, so
```

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -2074,6 +2074,72 @@ import {MyInput, MyInputForm} from './value_accessor_integration_spec';
+
+           tick(100);
+
+           // Now, the form is valid because its own asynchronous validation has resolved
+           // successfully (because ) the form control value `angular` includes the string`angular`
```
```suggestion
           // Now, the form is valid because its own asynchronous validation has resolved
           // successfully, because the form control value `angular` includes the `angular` string
```

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -2383,17 +2449,34 @@ import {MyInput, MyInputForm} from './value_accessor_integration_spec';
     return promise;
   };
 }
+function uniqLoginAsyncValidator(expectedValue: string, timeout: number = 0) {
```
May be add a line above the `uniqLoginAsyncValidator` function to visually separate it from the previous one.

AndrewKushnir reviewed (approved):
Hi @benbraou, thanks for applying additional changes and adding extra tests 👍 

I've looked at the changes again and left a few minor comments, mostly related to formatting (extra/missing line, etc). The changes LGTM, so I'm approving this PR. I will add the "cleanup" label that I will remove once the mentioned formatting comments are addressed.

Also an update after running tests in Google's codebase: all tests are passing ([external link](https://test.corp.google.com/ui#id=OCL:317598027:BASE:317730110:1592859640059:5ddd24c6&include_skipped=true)), which is great!

> I have just one remark regarding this new test that I added:

I looked at the test and at the comment, everything looks reasonable to me, thanks for adding a description of this situation in the test itself.

Thank you.

benbraou commented:
Hi @AndrewKushnir, I have addressed the remaining comments. 
Thank you for the review and your time!

AndrewKushnir commented:
Thanks for addressing the feedback @benbraou.

FYI, I've rebased your PR on top of the current master, since CI was "red" (since PR was not rebased recently).

As I mentioned in my previous message, running tests in Google's codebase didn't indicate any issues (including a global run in Google's codebase). Also given the fact that this is a bug fix with low risk, I'm removing the "breaking changes" flag. In any case, since the behavior of the framework changes slightly (to become more accurate), we would merge it to master only and it will be released as a part of the next minor version (v10.1.0).

Thank you.

benbraou commented:
Great news. Thank you !

AndrewKushnir commented:
Hi @benbraou, just wanted to let you know that your PR is now merged to master and the changes should be available in `v10.1.0` release of Angular 🎉 

Please let me know if you might be interested in doing some further refactoring around async validators in the tests (to reuse one async validator across multiple test cases). Thank you.

benbraou commented:
@AndrewKushnir thank you! Yes, I am interested to continuing the refactoring in tests. Can you please let me know if you want me to create a new issue for this?
Thank you

AndrewKushnir commented:
Hi @benbraou, thanks for the reply.

Yes, I think creating a new ticket would be great, thanks.

I was thinking if we can extract common validators code into a new `packages/forms/test/util.ts` file (we can store more functions there later). Some validators that are used across many tests (may be you can find more):

- function asyncValidator
- function asyncValidatorReturningObservable
- may be `simpleAsyncValidator` added in this test can use some common code from `asyncValidator`

Please let me know if any additional information is required, I'll be glad to help.

Thank you.
